### PR TITLE
[RFR] SUI:Automate chargeback test for service with multiple VMs

### PR DESCRIPTION
**Pull request by nachandr**
## Purpose or Intent
1)Automated chargeback test for a service with multiple VMs
2)Parametrized the order_service fixture on vm_count
3)Fixed the test_vm_console test so that it passes a key value pair to the vm_count parameter to the order_service fixture

### PRT Run
{{ pytest: --long-running cfme/tests/ssui/test_ssui_dashboard.py::test_service_chargeback_multiple_vms cfme/tests/ssui/test_ssui_myservice.py::test_vm_console }}

PRT Results:
510 - PASS
511 - PASS

### CODE REVIEW

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 1d0f602692d3ef20be7302f469d7de70d0213575

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 48cc333b60795cdf330edfdb1f6b26ab8bb1305e

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 6547c1abe27e50a1812a4f75623e507fdf246d20

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/ssui/test_ssui_dashboard.py [hunk 1]
```diff
@@ -235,6 +235,26 @@ def test_monthly_charges(appliance, has_no_providers_modscope, setup_provider, c
         assert monthly_charges != '$0'
 
 
+@pytest.mark.long_running
+@pytest.mark.parametrize('context', [ViaSSUI])
+@pytest.mark.parametrize('order_service', ['vm_count_1', 'vm_count_2'], indirect=True)
```
Comment (izapolsk) [inline cfme/tests/ssui/test_ssui_dashboard.py]: this doesn't look optimal. below is example how you can implement this:

```
import pytest


@pytest.fixture
def my_lovely_fixture(request):
    if isinstance(request.param, list):
        print(request.param[0], request.param[1])
    if isinstance(request.param, dict):
        print(request.param)
    yield request.param


@pytest.mark.parametrize('my_lovely_fixture', [{'console_test': True, 'vm_count': 1},
                                               {'console_test': False, 'vm_count': 2}],
                         indirect=True, ids=["vm-count1", "vm-count2"])
def test_my_lovely_test(my_lovely_fixture):
    assert True


@pytest.mark.parametrize('my_lovely_fixture', [[True, 1], [False, 2]],
                         indirect=True, ids=["vm-count1", "vm-count2"])
def test_my_lovely_test2(my_lovely_fixture):
    assert True
```

output:
```
test_params.py::test_my_lovely_test[vm-count1] {'console_test': True, 'vm_count': 1}
PASSED                    [ 25%]
test_params.py::test_my_lovely_test[vm-count2] {'console_test': False, 'vm_count': 2}
PASSED                    [ 50%]
test_params.py::test_my_lovely_test2[vm-count1] True 1
PASSED                   [ 75%]
test_params.py::test_my_lovely_test2[vm-count2] False 2
PASSED                   [100%]

============================== 4 passed in 0.02s ===============================
```

**Comment by dajoRH** [comment]
I detected some fixture changes in commit cc2c62c13247a45a0f8caab7605376a154c401c6

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 8012880243d2a8a3ee90699cfb64330efe711a97

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/ssui/test_ssui_myservice.py [hunk 2]
```diff
@@ -128,7 +128,7 @@ def _finalize():
 
 @pytest.mark.long_running
 @pytest.mark.parametrize('context', [ViaSSUI])
-@pytest.mark.parametrize('order_service', [['console_test']], indirect=True)
+@pytest.mark.parametrize('order_service', [{'console_test': True}], indirect=True)
```
Comment (mshriver) [inline cfme/tests/ssui/test_ssui_myservice.py]: Looks like this was the only other use of it, good.

File: cfme/tests/ssui/test_ssui_dashboard.py [hunk 3]
```diff
@@ -235,6 +235,27 @@ def test_monthly_charges(appliance, has_no_providers_modscope, setup_provider, c
...
+    Polarion:
+        assignee: nachandr
+        casecomponent: SelfServiceUI
+        caseimportance: high
+        initialEstimate: 1/2h
+    """
+    with appliance.context.use(context):
+        dashboard = Dashboard(appliance)
+        monthly_charges = dashboard.monthly_charges()
+        logger.info('Monthly charges is {}'.format(monthly_charges))
```
Comment (mshriver) [inline cfme/tests/ssui/test_ssui_dashboard.py]: When logging, please use `logger.info('The string %s', the_var)` syntax

File: cfme/fixtures/service_fixtures.py [hunk 4]
```diff
@@ -173,15 +177,17 @@ def order_service(appliance, provider, provisioning, dialog, catalog, request):
...
     yield catalog_item
     service = MyService(appliance, catalog_item.name)
     if service.exists:
         service.delete()
-    vm_name = '{}0001'.format(catalog_item.prov_data['catalog']['vm_name'])
-    vm = appliance.collections.infra_vms.instantiate(vm_name, provider)
-    vm.cleanup_on_provider()
+    name = catalog_item.prov_data['catalog']['vm_name']
+    for i in range(vm_count):
+        vm_name = f'{name}000'f'{i+1}'
```
Comment (mshriver) [inline cfme/fixtures/service_fixtures.py]: Is there a reason there are two f-strings here?

**Comment by mshriver** [comment]
Looks good @nachandr, just two small comments from me.

**Comment by dajoRH** [comment]
I detected some fixture changes in commit d43e1ad5a05caae723d77ebfc0f4f2b8e75a29ab

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 02911de61bda095b4ca77da9d20438ae966b6860

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

The *local* fixture **`run_service_chargeback_report`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/fixtures/service_fixtures.py [hunk 5]
```diff
@@ -159,12 +160,15 @@ def create_catalog_item(appliance, provider, provisioning, dialog, catalog,
...
 @pytest.fixture
 def order_service(appliance, provider, provisioning, dialog, catalog, request):
+    # BZ 1646333 - Delete this request button is not shown in service Request details page
+    # The above BZ got closed because of  INSUFFICIENT_DATA.
     """ Orders service once the catalog item is created"""
-
     if hasattr(request, 'param'):
-        param = request.param
+        vm_count = request.param['vm_count'] if 'vm_count' in request.param else '1'
+        console_test = request.param['console_test'] if 'console_test' in request.param else False
```
Comment (kedark3) [inline cfme/fixtures/service_fixtures.py]: Can you please elaborate why `console_test` is defaulting to `False` over `None`?

File: cfme/fixtures/service_fixtures.py [hunk 6]
```diff
@@ -173,15 +177,17 @@ def order_service(appliance, provider, provisioning, dialog, catalog, request):
     assert provision_request.is_succeeded()
     if provision_request.exists():
         provision_request.wait_for_request()
-        if not BZ(1646333, forced_streams=['5.10']).blocks:
+        if not BZ(1775779, forced_streams=['5.10', '5.11']).blocks:
```
Comment (kedark3) [inline cfme/fixtures/service_fixtures.py]: :+1:

File: cfme/tests/ssui/test_ssui_myservice.py (hunk 2 above)
Comment (kedark3) [reply cfme/tests/ssui/test_ssui_myservice.py]: @nachandr @mshriver do we want to test Console Tests in the PRT as well for this change?

File: cfme/fixtures/service_fixtures.py (hunk 5 above)
Comment (nachandr) [reply cfme/fixtures/service_fixtures.py]: The existing code already specifies that the default value of the 'console_test' keyword argument is False.
Am I missing something ?

```suggestion
        def create_catalog_item(appliance, provider, provisioning, dialog, catalog,
        console_test=False):
```

Comment (kedark3) [reply cfme/fixtures/service_fixtures.py]: Ah right, I see it now. Thanks for pointing out.

File: cfme/fixtures/service_fixtures.py [hunk 7]
```diff
@@ -80,7 +80,7 @@ def generic_catalog_item(request, appliance, dialog_modscope, catalog_modscope):
 
 
 def create_catalog_item(appliance, provider, provisioning, dialog, catalog,
-        console_test=False):
+         vm_count='1', console_test=False):
```
Comment (john-dupuy) [inline cfme/fixtures/service_fixtures.py]: Why does this default to a string? I would expect it to be an `int`

File: cfme/tests/ssui/test_ssui_dashboard.py (hunk 3 above)
Comment (nachandr) [reply cfme/tests/ssui/test_ssui_dashboard.py]: Done

File: cfme/fixtures/service_fixtures.py (hunk 4 above)
Comment (nachandr) [reply cfme/fixtures/service_fixtures.py]: That's an oversight. Fixed it.

File: cfme/fixtures/service_fixtures.py (hunk 7 above)
Comment (nachandr) [reply cfme/fixtures/service_fixtures.py]: @john-dupuy , The Catalog item creation form expects the value assigned to VM count to be a string.

File: cfme/tests/ssui/test_ssui_myservice.py (hunk 2 above)
Comment (mshriver) [reply cfme/tests/ssui/test_ssui_myservice.py]: PRT directive has been updated, thanks @nachandr

File: cfme/fixtures/service_fixtures.py (hunk 6 above)
Comment (mshriver) [reply cfme/fixtures/service_fixtures.py]: Don't we want to actually work around this BZ, not just skip the remove_request call?

If I understand the BZ correctly, we can still remove, we just need to navigate away first?

**Comment by dajoRH** [comment]
I detected some fixture changes in commit d24eb1ab92012951910100ab3155ff6dadef3058

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

The *local* fixture **`run_service_chargeback_report`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/fixtures/service_fixtures.py (hunk 6 above)
Comment (kedark3) [reply cfme/fixtures/service_fixtures.py]: I am actually proposing this change in the fixture: 
https://github.com/ManageIQ/integration_tests/pull/9663/files#diff-c43bfe6be31a2fdeb6a2768d21c9afe5R175

We don't care if UI/API is used, all we care is to remove the request. Hence using API method. And it works well, I have tested it.

**Comment by john-dupuy** [comment]
Can you check the PRT failures @nachandr?

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 0b7a7dfa032a43f5ddb0b7cacb40566bc53daa81

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

The *local* fixture **`run_service_chargeback_report`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/fixtures/service_fixtures.py (hunk 6 above)
Comment (nachandr) [reply cfme/fixtures/service_fixtures.py]: Yes, that works .

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 079f7899167fca828ff04e46a9a71f81d0689fa5

The *global* fixture **`order_service`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**
    - *test_total_services*
- **cfme/tests/ssui/test_ssui_myservice.py**
    - *test_myservice_crud*
- **cfme/tests/ssui/test_ssui_service_catalogs.py**

The *local* fixture **`run_service_chargeback_report`** is used in the following files:
- **cfme/tests/ssui/test_ssui_dashboard.py**

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by nachandr** [comment]
The test_vm_console tests failed on both 510 and 511.

Based on my initial investigation, the failures are because of these issues:
1)Element locator for console_type needs to be updated for 511.
2)It looks like VIEW in the LaunchVMConsole Class needs to be updated.

@kedark3 , could you review the PRT results for test_vm_console and let me know if the above issues would be fixed by https://github.com/ManageIQ/integration_tests/pull/9663/  ?

**Pull request merged**